### PR TITLE
Improve redux i in react

### PR DESCRIPTION
**Pull request by kapnobatai137**
export selectors with related reducers:
- insight was too advanced and it was moved to archived
- it will be re-used in a later workout

1. Use Redux with React:
- added example of how to make the store available
- re-wrote PQ

2. Use Redux with React (2):
- added example of connecting a component
- explained what the mapStateToProps function is
- explained what connect() does
- explained the process of connecting a component
- wrote new PQ

3. The Redux store:
- added explanation of what a reducer is
- added example of simple reducer
- added example of creating a store using the reducer

5. Your first reducer:
- added example of creating initial state and passing to reducer
- added example of an action object
- added example of reducer using the switch statement
- added example of dispatching actions in the component
- wrote PQ

### CODE REVIEW

**Comment by enki-hq** [comment]
### Insight previews:
- [ ] [enkidevs/curriculum/.archived/javascript/react/redux-i/export-selectors-with-related-reducers.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/.archived/javascript/react/redux-i/export-selectors-with-related-reducers.md?ref=cc5878e73aa0bf639c20555a6d26508993f1e70a)
- [ ] [enkidevs/curriculum/javascript/react/redux-i/the-redux-store.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/redux-i/the-redux-store.md?ref=cc5878e73aa0bf639c20555a6d26508993f1e70a)
- [ ] [enkidevs/curriculum/javascript/react/redux-i/understanding-pure-impure-functions.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/redux-i/understanding-pure-impure-functions.md?ref=cc5878e73aa0bf639c20555a6d26508993f1e70a)
- [ ] [enkidevs/curriculum/javascript/react/redux-i/use-redux-with-react-2.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/redux-i/use-redux-with-react-2.md?ref=cc5878e73aa0bf639c20555a6d26508993f1e70a)
- [ ] [enkidevs/curriculum/javascript/react/redux-i/use-redux-with-react.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/redux-i/use-redux-with-react.md?ref=cc5878e73aa0bf639c20555a6d26508993f1e70a)
- [ ] [enkidevs/curriculum/javascript/react/redux-i/your-first-reducer.md](https://app.enkipro.com/#/public/preview-insight/enkidevs/curriculum/javascript/react/redux-i/your-first-reducer.md?ref=cc5878e73aa0bf639c20555a6d26508993f1e70a)

File: javascript/react/redux-i/the-redux-store.md [hunk 1]
```diff
@@ -21,35 +23,60 @@ aspects:
...
 
 The **Redux** store is simply an object with few *methods* that holds the whole *state tree* of your application. The only way to change this state is by `dispatching` an action to the store.
 
-To create your store you need to pass your main `reducer` to the `createStore` function :
-```javascript
-import { createStore } from 'redux';
-let store = createStore(reducer, initState);
+Although we have connected our component in the previous insight, we still need to create our main `reducer`:
+
+```jsx
```
Comment (nem035) [inline javascript/react/redux-i/the-redux-store.md]: no jsx code here

File: javascript/react/redux-i/the-redux-store.md [hunk 2]
```diff
@@ -59,7 +86,7 @@ store.replaceReducer(newReducer);
 
 Create a Redux store and get the current state of the application:
 
-```
+```jsx
```
Comment (nem035) [inline javascript/react/redux-i/the-redux-store.md]: this is only JS too

File: javascript/react/redux-i/understanding-pure-impure-functions.md [hunk 3]
```diff
@@ -29,18 +29,18 @@ Example of a simple pure function:
...
 **Pure** functions must always return a new object and not modify the argument. For example, if we want to decrement every element of an `array`, the **pure** way to do it's by creating a new array with the same elements, but decremented.
 
-On the contrary, **impure** functions don't have a predictable behaviour as they might call the network, the database, modify the **DOM** or even alter the value/s passed to them.
+On the contrary, **impure** functions don't have a predictable behavior as they might call the network, the database, modify the **DOM** or even alter the value/s passed to them.
 
 Example:
 ```javascript
 function doubleImpure(x) {
-  x= 2 * x;
+  x = 2 * x;
```
Comment (nem035) [inline javascript/react/redux-i/understanding-pure-impure-functions.md]: hmmmm this isn't really impure. The changes do not propagate outside `doubleImpure` function.

These are better examples for pure vs impure:

```js
function doublePure(x) {
  // do not change x.
  // instead return a new object
  // with the change applied to it.
  return { ...x, num: x.num * 2 };
}

function doubleImpure(x) {
  // change the same object
  // this mutation affects the object
  // outside the lifetime of doubleImpure
  // which is not "pure"
  x.num = x.num * 2;
  return x;
}
```

File: javascript/react/redux-i/understanding-pure-impure-functions.md [hunk 4]
```diff
@@ -52,12 +52,12 @@ Are the following functions pure or impure?
 
 ```javascript
 function a(x) {
-  x = 'enki'
-  return x
+  x = "enki";
```
Comment (nem035) [inline javascript/react/redux-i/understanding-pure-impure-functions.md]: As mentioned above, this isn't impure since it's not a mutation that exists beyond the lifetime of `a`.

File: javascript/react/redux-i/use-redux-with-react-2.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  - introduction
+
+---
+
+# Use Redux with React (2)
+
+---
+## Content
+
+Now, looking in `\Counter.js`, we must connect the component to our `store`:
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react-2.md]:
```suggestion
Now, looking in `Counter.js`, we must connect the component to our `store`:
```

File: javascript/react/redux-i/use-redux-with-react-2.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    count: state.count
+  };
+}
+
+export default connect(mapStateToProps)(
+  Counter
+);
+```
+
+Let's look at this code step-by-step. We want to connect our `<Counter />` component to our `store`, and to do this we have to make use of the `connect()` method. What this method does is that it takes our component as input, and returns a **new** connected component class that wraps our `<Counter />` component. Notice how you pass the `mapStateToProps` and `Counter` to the `connect` method. This is because `connect()` is a higher order function, which basically means that it returns another function that takes `Counter` as argument.
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react-2.md]:
```suggestion
Let's look at this code step-by-step. We want to connect our `<Counter />` component to our `store`, and to do this we have to make use of the `connect()` method. What this method does is that it takes our component as input, and returns a **new** connected component class that wraps our `<Counter />` component. Notice how you pass the `mapStateToProps` and `Counter` to the `connect` method. This is because `connect()` is a higher order function, which basically means that it returns another function that takes `Counter` as an argument.
```

File: javascript/react/redux-i/use-redux-with-react-2.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+The last piece of the puzzle is the `mapStateToProps` function which does exactly what the name suggests. The `<Counter />` component has `count: 0` as a state, and the `mapStateToProps` function will create a new object where the key represents the name of the `prop`, in our case `count`, and the value will be extracted from `this.state.count`. Now, when you use `connect()` to create a new connected component, the method will pass the entire state of the `<Count />` component to the `mapStateToProps` function which in turn feeds your returned object as `prop`s to they new **connected component**.
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react-2.md]:
```suggestion
The last piece of the puzzle is the `mapStateToProps` function which does exactly what the name suggests. The `<Counter />` component has `count: 0` as a state, and the `mapStateToProps` function will create a new object where the key represents the name of the `prop`, in our case `count`, and the value will be extracted from `state.count`. Now, when you use `connect()` to create a new connected component, the method will pass the entire state of the `<Count />` component to the `mapStateToProps` function which in turn feeds your returned object as `prop`s to the new **connected component**.
```

File: javascript/react/redux-i/use-redux-with-react-2.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+* `(mapStateToProps)`
+* `(Counter)`
+* `(mapPropsToState)`
+* `(mapProps)`
+* `(mapState)`
+
+---
+## Revision
+
+Connect the `<Counter />` component to the Redux `store`:
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react-2.md]: why the identical question for both practice and revision?

File: javascript/react/redux-i/use-redux-with-react.md [hunk 9]
```diff
@@ -27,67 +25,85 @@ aspects:
...
 ```bash
 npm install --save react-redux
 ```
+
 The general paradigm is that only top level components (e.g. route handlers) should be aware of Redux.
 
-To turn a presentational container into a component, the `connect` function must be imported:
-```javascript
-import { connect } from 'react-redux'
+Let's say that you have a `<Counter />` component found in `\Counter.js`. The first step in using Redux is making the `store` available to your app. This is done by wrapping the main component (usually found in `\App.js`) with the `<Provider />` API, which is provided by React-Redux.
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react.md]:
```suggestion
Let's say that you have a `<Counter />` component found in `Counter.js`. The first step in using Redux is making the `store` available to your app. This is done by wrapping the main component (usually found in `App.js`) with the `<Provider />` API, which is provided by React-Redux.
```

File: javascript/react/redux-i/use-redux-with-react.md [hunk 10]
```diff
@@ -27,67 +25,85 @@ aspects:
...
-Inject `dispatch` and don't listen to the store:
-```javascript
-export default connect()(myReducer)
+ReactDOM.render(
+  <App />,
+  document.getElementById("root")
+);
 ```
 
+Note that you must also pass a `store` attribute to the `<Provider />` wrapper, which takes a value of `{store}`.
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react.md]:
```suggestion
Note that you must also pass a `store` attribute to the `<Provider />` wrapper.
```

File: javascript/react/redux-i/your-first-reducer.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+aspects:
+  - introduction
+
+---
+# Your first reducer
+
+---
+## Content
+
+In a previous insight we have created a simple `reducer` that looked something like this:
```
Comment (nem035) [inline javascript/react/redux-i/your-first-reducer.md]: Why `reducer` and not reducer? It's not code, right?

File: javascript/react/redux-i/your-first-reducer.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+function reducer(
+  state = initialState,
+  action
+) {
+  return state;
+}
+```
+
+You might be wondering what an `action` is, and the answer to that is: a plain object that has a property called `type`:
```
Comment (nem035) [inline javascript/react/redux-i/your-first-reducer.md]:
```suggestion
You might be wondering what an `action` is, and the answer to that is: a plain object that has a property called `type` and (optionally) other properties specific to that action type:
```

File: javascript/react/redux-i/your-first-reducer.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+```jsx
+{
+  type: "add an item",
+  item: "banana"
+}
+```
+
+For our component, we will want to increase, decrease or reset the counter. To do this, we would write our `reducer` like this:
```
Comment (nem035) [inline javascript/react/redux-i/your-first-reducer.md]:
```suggestion
For our component, we will want to increase, decrease or reset the counter. To do this, we would write our `reducer` such that it inspects the `type` of an action received and returns the new updated state for a particular action:
```

File: javascript/react/redux-i/understanding-pure-impure-functions.md (hunk 3 above)
Comment (kapnobatai137) [reply javascript/react/redux-i/understanding-pure-impure-functions.md]: I've changed the examples to the ones suggested. I've also added the **pure** vs **impure** example in which we show predictable and consistent output for the same input.

File: javascript/react/redux-i/understanding-pure-impure-functions.md (hunk 4 above)
Comment (kapnobatai137) [reply javascript/react/redux-i/understanding-pure-impure-functions.md]: changed the question to

```
function a(x) {
  return x * x;
}

function b(x) {
  return Math.random() * x;
}
```

File: javascript/react/redux-i/use-redux-with-react-2.md (hunk 8 above)
Comment (kapnobatai137) [reply javascript/react/redux-i/use-redux-with-react-2.md]: no other excuse for that other than being lazy. wrote a new RQ which is:

Finish the following code so that you can connect the component to the Redux `store`:

```js
function ???(state) {
  ??? {
    name: ???.name,
    age: ???.age
  };
}
```

* `mapStateToProps`
* `return`
* `state`
* `state`
* `props`
* `props`
* `mapPropsToState`

File: javascript/react/redux-i/your-first-reducer.md (hunk 11 above)
Comment (kapnobatai137) [reply javascript/react/redux-i/your-first-reducer.md]: You are right, wanted to evidentiate it but I could've done that with bold. I've removed the code ticks from `reducer`.

**Comment by enki-hq** [comment]
Rule insight:code-width failed for files: 
  - javascript/react/redux-i/understanding-pure-impure-functions.md

File: javascript/react/redux-i/understanding-pure-impure-functions.md [hunk 14]
```diff
@@ -29,35 +29,60 @@ Example of a simple pure function:
...
+  return x * randomInteger();
+}
+```
+
+The **pure** function provides a predictable and consistent output for the same input, while the **impure** function does not.
+
 ---
 ## Practice
 
 Are the following functions pure or impure?
```
Comment (nem035) [inline javascript/react/redux-i/understanding-pure-impure-functions.md]: Maybe rephrase this one to explain in the answers that we want them to answer `a`, then `b`?

File: javascript/react/redux-i/use-redux-with-react-2.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    count: state.count
+  };
+}
+
+export default connect(mapStateToProps)(
+  Counter
+);
+```
+
+Let's look at this code step-by-step. We want to connect our `<Counter />` component to our `store`, and to do this we have to make use of the `connect()` method. What this method does is that it takes our component as input, and returns a **new** connected component class that wraps our `<Counter />` component. Notice how you pass the `mapStateToProps` and `Counter` to the `connect` method. This is because `connect()` is a higher order function, which basically means that it returns another function that takes `Counter` as an argument.
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react-2.md]: How about this?

```suggestion
Let's look at this code step-by-step. We want to connect our `<Counter />` component to our `store`, and to do this we have to make use of the `connect()` method. What this method does is that it takes our component as input, and returns a **new** connected component class that wraps our `<Counter />` component. Notice how you pass the `mapStateToProps` and `Counter` to the `connect` method. This is because `connect()` is a higher order function, which basically means it is a "function that returns a function" and the function `connect` returns in this case is one that it returns another function that takes `Counter` as an argument and returns a new `Counter` that is connected to the redux store.
```

File: javascript/react/redux-i/understanding-pure-impure-functions.md (hunk 14 above)
Comment (kapnobatai137) [reply javascript/react/redux-i/understanding-pure-impure-functions.md]: Changed to 

Consider these functions:

```javascript
function a(x) {
  return x * x;
}

function b(x) {
  return Math.random() * x;
}
```

Would you say function `a` is pure or impure? What about function `b`?

File: javascript/react/redux-i/use-redux-with-react-2.md [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    count: state.count
+  };
+}
+
+export default connect(mapStateToProps)(
+  Counter
+);
+```
+
+Let's look at this code step-by-step. We want to connect our `<Counter />` component to our `store`, and to do this we have to make use of the `connect()` method. What this method does is that it takes our component as input, and returns a **new** connected component class that wraps our `<Counter />` component. Notice how you pass the `mapStateToProps` and `Counter` to the `connect` method. This is because `connect()` is a higher order function, which basically means it is a "function that returns a function" and the function `connect` returns, in this case, is one that takes `Counter` as an argument and returns a new `Counter` that is connected to the redux store.
```
Comment (nem035) [inline javascript/react/redux-i/use-redux-with-react-2.md]: Just a tid bit more lol

```suggestion
Let's look at this code step-by-step. We want to connect our `<Counter />` component to our `store`, and to do this we have to make use of the `connect()` method. What this method does is that it takes our component as input, and returns a **new** connected component class that wraps our `<Counter />` component. Notice how you pass the `mapStateToProps` and `Counter` to the `connect` method, through separate function calls. This is because `connect()` is a higher order function, which basically means it is a "function that returns a function". The function that `connect` returns, in this case, is one that takes `Counter` as an argument and returns a new `Counter` that is connected to the redux store.
```
